PORT-10112 Comprehensive DORA Metrics guide (#1594)
port-labs/port-docs: docs/guides/all/setup-dora-metrics.md

- Link incidents to services to track the impact of failures.
- Aggregate metrics across incidents for better monitoring.

### Data Model Setup

Ensure that your **PagerDuty incident blueprint** is properly configured to map incidents to the correct services(gitHub). 
This includes defining the appropriate properties and relations for incidents.Follow this [PagerDuty Incident Blueprint Setup Link](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty#incident-blueprint-setup) to implement.

Add the following properties to capture incident resolution time and recovery time:

<details>
<summary><b>Additional properties for PagerDuty Incident Blueprint (click to expand)</b></summary>

```json showLineNumbers
 "resolvedAt": {
   "title": "Incident Resolution Time",
   "type": "string",
   "format": "date-time",
   "description": "The timestamp when the incident was resolved"
 },
 "recoveryTime": {
   "title": "Time to Recovery",
   "type": "number",
   "description": "The time (in minutes) between the incident being triggered and resolved"
 }

```
</details>

### Syncing Incidents with PagerDuty and Other Tools

> To sync incidents from **PagerDuty**, follow the steps in the [PagerDuty guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty). The guide provides detailed steps for setting up integrations to track incidents related to deployments.

Reviewer: @kodjomiles I believe the correct way to link to docs is without the base domain
```suggestion
To sync incidents from **PagerDuty**, follow the steps in the [PagerDuty guide](/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty). The guide provides detailed steps for setting up integrations to track incidents related to deployments.
```

- [ServiceNow](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/servicenow)
- [Statuspage](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/statuspage)

### Relating Incident to services
Add this relationship to the **Incident blueprint** to link incidents to GitHub repository (service):

<details> 
<summary><b>PagerDuty Incident Relations</b> (click to expand)</summary>

 ```json showLineNumbers
 "gitHubRepository": {
   "title": "Github Service",
   "target": "service",
   "required": false,
   "many": false
 }

```
</details>

Update the mapping config to pagerduty incident [data source](https://app.getport.io/settings/data-sources):
<details>
<summary><b>Incident mapping config (click to expand)</b></summary>

```yaml showLineNumbers
   - kind: incidents
     selector:
        query: 'true'
        ...: # Add other selectors as needed
     port:
        entity:
           mappings:
              identifier: .id | tostring
              title: .title
              blueprint: '"pagerdutyIncident"'
              properties:
                 status: .status

>                  url: .self

Reviewer: @kodjomiles Aren't we missing with the definition of how to populate recoveryTime field?

Author: You are right, resolvedAt and recoveryTime, thought we had covered it earlier

```json showLineNumbers

"deployment_frequency": {
   "title": "Monthly Deployment Frequency",
   "icon": "DeploymentIcon",
   "type": "number",
   "target": "deployment",
   "query": {
      "combinator": "and",
      "rules": [
         {
            "property": "deploymentStatus",
            "operator": "=",
            "value": "Success"
         }
      ]
   },
   "calculationSpec": {
      "func": "average",
      "averageOf": "month",
      "measureTimeBy": "$createdAt",
      "calculationBy": "entities"
   }
}

```
</details>
</TabItem>

<TabItem value="clt" label="Change Lead Time">
Add this aggregation property to calculate the lead time for changes:
<details>
<summary><b>Lead Time for Change (click to expand)</b></summary>

```json showLineNumbers

"lead_time_for_change": {
   "title": "Lead Time for Change",
   "icon": "DefaultProperty",
   "type": "number",
   "target": "githubPullRequest",
   "query": {
     "combinator": "and",
     "rules": [
       {
         "property": "status",
         "operator": "=",
         "value": "merged"
       }
     ]
   },
   "calculationSpec": {
     "func": "average",
     "averageOf": "total",

>      "property": "leadTimeDays",

Reviewer: ```suggestion
     "property": "leadTimeHours",
```

Right? you changed the original prop name in the blueprint

Author: yeah

<details>
<summary><b>CFR calculation property (click to expand)</b></summary>
    
```json showLineNumbers
      "changeFailureRate": {
        "title": "Change Failure Rate",
        "calculation": "(.properties.resolved_incidents / .properties.total_deployments) * 100",
        "type": "number"
      }
     
```
</details>

</TabItem>

<TabItem value="mttr" label="Mean Time to Recovery">
Add this aggregation property to calculate the MTTR:

<details>
<summary><b>Mean Time to Recovery (click to expand)</b></summary>

```json showLineNumbers

   "mean_time_to_recovery": {
      "title": "Mean Time to Recovery",
      "icon": "DefaultProperty",
      "type": "number",
      "target": "pagerdutyIncident",
      "query": {
         "combinator": "and",
         "rules": [
            {
               "property": "status",
               "operator": "=",
               "value": "resolved"
            }
         ]
      },
      "calculationSpec": {
         "func": "average",
         "averageOf": "month",
         "property": "recoveryTime",
         "measureTimeBy": "$createdAt",
         "calculationBy": "property"
      }
   }

```

</details>

</TabItem>

</Tabs>

> 

Reviewer: @kodjomiles consider adding a note here explaining that at this point you can already go to each service and see the aggregation data. HOWEVER, it will be based on new data only (and you can link to the AGGREGATION PROP page to more information on how it works)

---
sidebar_position: 10
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

# Track DORA metrics

This guide is designed
to help you
implement and track [DevOps Research and Assessment (DORA) metrics](https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance) within your organization in Port.  

> DORA Metrics are a set of key performance indicators

Reviewer 2: ```suggestion

DORA Metrics are a set of key performance indicators
```

that measure the effectiveness and efficiency of your software development and delivery process. 
By tracking these metrics,
you can identify areas for improvement and ensure that your team is delivering high-quality software efficiently.
This guide will cover the four key metrics: **deployment frequency**, **lead time**, **change failure rate**, and **mean time to recovery**.

### Prerequisites
- Complete the [Port onboarding process](https://docs.getport.io/quickstart).

> - Access a GitHub repository (e.g., port-dora-metrics) to sync your pull requests or deployments. You can follow the [GitHub integration guide](/build-your-software-catalog/sync-data-to-catalog/git/github) to ensure your deployments are tracked in Port.

Reviewer 2: If a user completes the onboarding using GitHub, they will already have a repo that is synced with Port.

Instead of this line, let's split it to 2 lines:
The first should be something like "While this guide demonstrates an implementation using GitHub, other Git providers can be used as well."
And the second should be something like "Access to a GitHub repository that is connected to Port via the onboarding process."

:::

</TabItem>

<TabItem value="github-deployment" label="GitHub Deployment">

GitHub deployments can be tracked by mapping repository releases and tags to deployment entities in Port.
These repositories can hold critical information related to service versions, commits, and releases.

Here’s how you can implement this:

1. **Add the configuration below** to the [data sources page](https://app.getport.io/settings/data-sources) in your Port portal, and select your GitHub integration.

<details>
<summary><b>Releases and Tags config (click to expand)</b></summary>

```yaml showLineNumbers
- kind: workflow-run
   selector:
   query: .head_branch == 'main'
   port:
      entity:
         mappings:
            identifier: .head_repository.name + '-' + (.run_number|tostring)
            title: .head_repository.name + " Deployment on workflow"
            blueprint: '"deployment"'
            properties:
               environment: '"Production"'
               createdAt: .created_at
               deploymentStatus: (.conclusion | ascii_upcase[0:1] + .[1:])
            relations:
               service: .head_repository.name

> 

Reviewer 3: Why is the gitlab releases and tags using the same mapping as the Deployment via Workflow Runs?

Reviewer 3: https://pr-1619.d2ngvl90zqbob8.amplifyapp.com/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples#map-repositories-repository-releases-and-tags

Author: It's supposed to be for the release tags not the workflow, it's a mistake let me replace it

  "identifier": "custom-deployment-1234",
  "title": "Custom Deployment for ID 1234",
  "properties": {
    "environment": "Production",
    "createdAt": "2024-09-01T12:00:00Z",
    "deploymentStatus": "Success"
  },
  "relations": {
    "service": {
      "combinator": "and",
      "rules": [
        {
          "property": "$title",
          "operator": "=",
          "value": "Custom-Service-Name"
        }
      ]
    }
  }
}'
```
</details>

:::tip Custom api service relation
we use the **search relation** entity to map the deployment to the correct service based on the service's `$title`.
To learn more about using search relations,
see [our documentation on Mapping Relations
Using Search Queries](/build-your-software-catalog/customize-integrations/configure-mapping/#mapping-relations-using-search-queries).
You can change the query to match the naming convention in your organization
and remember to replace `$YOUR_PORT_API_TOKEN` with your actual API token from port
:::

By using this approach, you can ensure any deployment system is integrated into Port,
giving you full flexibility across all your deployment workflows.

</TabItem>

</Tabs>

<br/>

### Monorepo Tracking

> Port does not offer a built-in solution specifically tailored for monorepos.

Reviewer 2: ```suggestion
```

> However, by using **custom integrations**, you can track services or components within a monorepo effectively.

Reviewer 2: ```suggestion
By using **custom integrations**, you can effectively track services or components within a monorepo.
```

      "target": "pagerdutyIncident",
      "query": {
         "combinator": "and",
         "rules": [
            {
               "property": "status",
               "operator": "=",
               "value": "resolved"
            }
         ]
      },
      "calculationSpec": {
         "func": "average",
         "averageOf": "month",
         "property": "recoveryTime",
         "measureTimeBy": "$createdAt",
         "calculationBy": "property"
      }
   }

```

</details>

</TabItem>

</Tabs>

:::note Aggregation Data Availability
At this point, you can already visit each service to view the aggregated DORA metrics. However, note that the aggregation data will only be calculated based on newly ingested data moving forward.
click [here](/build-your-software-catalog/customize-integrations/configure-data-model/setup-blueprint/properties/aggregation-property) for more details on aggregation properties.
:::

## Visualization
By leveraging Port's Dashboards, you can create custom dashboards to track the metrics and monitor your team's performance over time.

### Dashboard Setup
1. Go to your [software catalog](https://app.getport.io/organization/catalog).

> 2. Click on the `+ New` button in the left sidebar

Reviewer 2: Add a period at the end of each item in a list

:::

## Metrics

We will now aggregate the **DORA metrics** at the **service level**, allowing us to track metrics such as **Deployment Frequency**, **Change Lead Time**, **Change Failure Rate (CFR)**, and **Mean Time to Recovery (MTTR)** for each service.

:::note Metrics Aggregation
If you want to track these metrics at higher levels, such as **team** or **domain**, makes sure the appropriate **team** or **domain** blueprints exist, and that they have relationships defined with the **service** blueprint. Then, you can apply the aggregation properties for these higher hierarchies, similar to how we are doing for the **service** blueprint below.
:::

###  Aggregation

The metrics in this guide are aggregated monthly. However, you can easily switch the timeframes to weekly, hourly, etc., based on your requirements.

<Tabs groupId="metrics" defaultValue="df" values={[
{ label: 'Deployment Frequency', value: 'df' },
{ label: 'Change Lead Time', value: 'clt' },
{ label: 'Change Failure Rate', value: 'cfr' },
{ label: 'Mean Time to Recovery', value: 'mttr' }
]}>

> <TabItem value="df" label="Deployment Frequency">

Reviewer: @kodjomiles I'm not sure it's entirely clear WHERE to add the aggregation/calcluation properties. You wrote above "service" but it wasn't `service`, rather just the text.

Consider explaining here how to actually reach the data model, that you need to select the service, update the JSON, and add it within the aggregation/calculation..
I think adding a screenshot here could help

Author: I have added a tip on how to get there. Look at it and let me know if I should still add a screenshot

Reviewer: @kodjomiles I don't see the tip, can you point me in the right direction?

Author: <img width="1023" alt="image" src="https://github.com/user-attachments/assets/64d372ff-68c7-41ba-8225-53b70fbbf417">
right here

Reviewer: Missed it, perfect

<details>
<summary><b>Lead Time for Change (click to expand)</b></summary>

```json showLineNumbers

"lead_time_for_change": {
   "title": "Lead Time for Change",
   "icon": "DefaultProperty",
   "type": "number",
   "target": "githubPullRequest",
   "query": {
     "combinator": "and",
     "rules": [
       {
         "property": "status",
         "operator": "=",
         "value": "merged"
       }
     ]
   },
   "calculationSpec": {
     "func": "average",
     "averageOf": "total",
     "property": "leadTimeHours",
     "measureTimeBy": "$createdAt",
     "calculationBy": "property"
   }
 }

```
</details>
</TabItem>

<TabItem value="cfr" label="Change Failure Rate">
Add the following to the aggregated property in service:
<details>
<summary><b>CFR aggregation property (click to expand)</b></summary>

```json showLineNumbers

  "resolved_incidents": {
    "title": "Resolved Incidents",
    "type": "number",
    "target": "pagerdutyIncident",
    "query": {
      "combinator": "and",
      "rules": [
        {
          "property": "status",
          "operator": "=",
          "value": "resolved"
        },
        {

>           "property": "incidentType",

Reviewer: @kodjomiles I didn't recognize this property. In my env I don't have it. Why do we need it?
The way I see it each incident should be counted

Author: true

Author: CFR is calculated based on the percentage of failed deployments that lead to incidents or issues that were eventually resolved.
counting only resolved incidents provides clarity on how many deployment-related incidents have been addressed, which is key to understanding recovery efforts and not just the raw number of incidents. right ?? and the unresolved incidents would indicate that the failures are still in progress, which wouldn’t accurately reflect the **recovery** or **success rate** of deployments.

Author: If I include **unresolved incidents**, doesn't that mean that the CFR might inflate because ongoing incidents don’t reflect the true end state (resolved or unresolved) of the failure.
Resolved incidents show how many incidents were fully dealt with and would be directly related to the overall health of the system, which is what CFR aims to measure, correct me if I am wrong

Reviewer: @kodjomiles you're not wrong but when calculating CFR you should count as you mentioned all failed deployments. If an incident is not resolved, you can still consider it a failed deployment (it's still an incident, just an ongoing one).
So the way I see it in this guide each incident is considered a failed deployment

Author: okay let me modify it then

2. Click on the `+ New` button in the left sidebar
3. Select **New dashboard**
4. Name the dashboard (e.g., DORA Metrics), choose an icon if desired, and click `Create`

This will create a new empty dashboard. Let's get ready-to-add widgets 

### Adding Widgets

<details>
<summary><b>Setup Deployment Frequency Widget</b></summary>

1. Click `+ Widget` and select **Number Chart**
2. Title: `Deployment Frequency - Monthly`, (add the rocket icon)
3. Select `Display single property` and choose **Service** as the **Blueprint**
4. Select an `Entity` and choose `Monthly Deployment Frequency` as the **Property**

   <img src="/img/guides/deploymentFrequencyChartDoraMetrics.png" width="50%"/>
   
5. Click `Save`

</details>

<details>
<summary><b>Setup MTTR Widget</b></summary>

1. Click `+ Widget` and select **Number Chart**
2. Title: `MTTR – Monthly Average (Seconds)`, (add the pagerduty icon)
3. Select `Display single property` and choose **Service** as the **Blueprint**
4. Select an `Entity` and choose `Mean Time to Recovery` as the **Property**

   <img src="/img/guides/mttrDoraMetricsChart.png" width="50%"/>
   
5. Click `Save`

</details>

<details>

> <summary><b>Setup Change Lead Time Widget</b></summary>

Reviewer: @kodjomiles there are few discrepencies here:

1. When I set it up, I see a different popup
![CleanShot 2024-09-21 at 23 06 28@2x](https://github.com/user-attachments/assets/94e4cb76-6cc4-4f18-a138-b2a2cb1b6c21)
2. I can't select entity, if we aim to show the dashboard per entity you should specify how to filter only a specific entity.
3. The title is "monthly", while in section 7 you with (e.g., days) and the screenshot you write "hour". Make sure the timeframe is clear and be opinionated (instead of e.g., we should write what you expect the user to do)

### Data model setup

1. Navigate to your [Port Builder](https://app.getport.io/settings/data-model) page.
2. Click the `+ Blueprint` button to create a new blueprint.
3. Name it `Deployment` and add the schema below:

<details>
<summary><b>Deployment blueprint (click to expand)</b></summary>

```json showLineNumbers
{
  "identifier": "deployment",
  "title": "Deployment",
  "icon": "Rocket",
  "schema": {
    "properties": {
      "createdAt": {
        "title": "Deployment Time",
        "type": "string",
        "format": "date-time",
        "description": "The timestamp when the deployment was triggered."
      },
      "environment": {
        "title": "Environment",
        "type": "string",
        "enum": [
          "Production",
          "Staging",
          "Testing"
        ],
        "description": "The environment where the deployment occurred."
      },
      "deploymentStatus": {
        "title": "Deployment Status",
        "type": "string",
        "enum": [
          "Success",
          "Failed"
        ],
        "description": "Indicates whether the deployment was successful or failed."
      }
    },
    "required": []
  },
  "mirrorProperties": {

>     "leadTimeHours": {

Reviewer: There's currently a bug with this approach, discussed in https://getport.slack.com/archives/C07ALEX911V/p1726725961241489?thread_ts=1726507073.753509&cid=C07ALEX911V
If the bug is fixed, no need to change. Otherwise, we should separate the "create blueprint" and "Add mirror property"

Author: this links seems to be broken

Reviewer: @kodjomiles I think you are not in this channel, but currently you can't copy paste the code you added. I confirmed the bug is fixed.

Author: but you can add them separately ?

Author: How come it worked for me ?

Reviewer: Separately it works. 
Try this:
Copy the blueprint from the doc
Change the relation name to one that doesn't exists (eg, pullRequest2)
Change the mirror property to that relation as well

See if you see an error

      "title": "Lead Time (Hours)",

>       "path": "pullRequest.leadTimeHours"

Reviewer: In the Github documentation the name of the field is `lead_time_days`
We should be consistent.
Also, I suggest adding here a note to make sure you followed the Github examples and that you defined this field. IF you havene't, LINK to where you can do it

Author: well now if they follow the link, it's a calculation property that can not be mirrored, that's why it's left out for now, I changed it to hours because that's what I saw in the notion docs, Let me change to days then. consistency across is great

Author: @Matanga1-2 should I change this to days ??

>     }

Reviewer: According to the Github guide, I need to create a calculation property on the lead time. I followed the example here
https://pr-1594.d2ngvl90zqbob8.amplifyapp.com/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples#map-repositories-and-pull-requests

But then I can't create an aggregation property of it. I think we discussed it.
What should be the approach here? the Github guide should be updated as well on how to create the required fields

Author: Yes please, the GitHub guide needs to be updated now to configure lead time as a property and not as a calculation property with the necessary mapping config added. I had that mapping in the code earlier but you suggested we remove it and update the GitHub docs instead

{label: "Custom API", value: "custom-api"}
]}>

<TabItem value="pr-merge" label="PR Merge">

One of the ways to track deployments is by monitoring when pull requests (PRs) are merged into a branch,
typically the **main**/**master** branch.
This is the **recommended** approach for tracking deployments and calculating lead time.

The lead time for these merges is calculated as the difference between when the PR was created and when it was merged.

**Example**:

- When a PR is merged, a **deployment entity** is created in Port to represent the deployment that took place.
- The **lead time** for that PR is calculated and added to the deployment as part of the blueprint.

Here’s how you can implement this:

1. **Add Pull Request blueprint, sample can be found [here](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github/examples/resource-mapping-examples/#map-repositories-and-pull-requests)**:

2. **Add the configuration below** to the [data sources page](https://app.getport.io/settings/data-sources) in your Port portal, and select your GitHub integration:

<details>
<summary><b>Deployment config (click to expand)</b></summary>

> ```yaml showLineNumbers

Reviewer: I did this step, and tested the mapping. I got this
![CleanShot 2024-09-19 at 09 15 39](https://github.com/user-attachments/assets/5c0d084c-d361-4183-bf09-2a1f44049c94)

It looks like it takes "Open" PR and considers it a successful deployment (without "Created at"). Not sure it's expected

Author: @Matanga1-2 I am certain that we need to add an additional filter to the query. since this is being tracked when the pr is merged then we need to add that the status should be merged too, so that we can have the deployment being created at mergedAt

:::

## Tracking Incidents

Incidents are essential for tracking key DORA metrics, including **Change Failure Rate (CFR)** and **Mean Time to Recovery (MTTR)**. Effective incident tracking reveals insights into how frequently deployments fail and how quickly teams can resolve issues. This section outlines how to:

- Use incidents to calculate **CFR** and **MTTR**.
- Link incidents to services to track the impact of failures.
- Aggregate metrics across incidents for better monitoring.

### Data Model Setup

Ensure that your **PagerDuty incident blueprint** is properly configured to map incidents to the correct services(gitHub). 
This includes defining the appropriate properties and relations for incidents.Follow this [PagerDuty Incident Blueprint Setup Link](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty#incident-blueprint-setup) to implement.

Add the following properties to capture incident resolution time and recovery time:

<details>
<summary><b>Additional properties for PagerDuty Incident Blueprint (click to expand)</b></summary>

```json
{

>   "properties": {

Reviewer: @kodjomiles it a bit weird to copy paste it as I have additional properties. Maybe there's a better way?

    }
  }
}
```
</details>

### Syncing Incidents with PagerDuty and Other Tools

To sync incidents from **PagerDuty**, follow the steps in the [PagerDuty guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty). The guide provides detailed steps for setting up integrations to track incidents related to deployments.

For other incident management tools, follow these respective guides:
- [OpsGenie](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/opsgenie)
- [FireHydrant](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/firehydrant)
- [ServiceNow](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/servicenow)
- [Statuspage](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/statuspage)

### Relating Incident to services
Add this relationship to the **Incident blueprint** to link incidents to GitHub repository (service):

<details> 
<summary><b>PagerDuty Incident Relations</b> (click to expand)</summary>

 ```json showLineNumbers

>   "relations": {

Reviewer: @kodjomiles Same here, is there a better approach to make it copy-paste? not a must

    "gitHubRepository": {
      "title": "Github Service",
      "target": "service",
      "required": false,
      "many": false
    }
  }

```
</details>

Add the mapping config to pagerduty incident [data source](https://app.getport.io/settings/data-sources):
<details>
<summary><b>Incident mapping config (click to expand)</b></summary>

> ```yaml showLineNumbers

Reviewer: @kodjomiles it took me a few seconds to understand where to paste it, and even then since yaml is sensitive it didn't work well.
Can you try to improve it, or at least show a screenshot to clarify where to put it?

   gitHubRepository:
   combinator: '"and"'
   rules:
     - property: '"$title"'
       operator: '"="'

>        value: .summary

Reviewer: ```suggestion
       value: .service.summary
```

   "total_deployments": {
      "title": "Total Monthly Deployment Frequency",
      "type": "number",
      "target": "deployment",
      "query": {
         "combinator": "and",
         "rules": [
            {
               "property": "deploymentStatus",
               "operator": "=",
               "value": "Success"
            }
         ]
      },
      "calculationSpec": {
         "func": "average",
         "averageOf": "month",
         "measureTimeBy": "$createdAt",
         "calculationBy": "entities"
      }
   }
}
```
</details>

Add this calculation property to calculate the cfr from the aggregated properties:
<details>
<summary><b>CFR calculation property (click to expand)</b></summary>
    
```json showLineNumbers
    {
      "changeFailureRate": {
        "title": "Change Failure Rate",
        "calculation": "(.properties.resolved_incidents / .properties.total_deployments) * 100",
        "type": "number"
      }
    }
     
```
</details>

</TabItem>

<TabItem value="mttr" label="Mean Time to Recovery">
Add this aggregation property to calculate the MTTR:

<details>
<summary><b>Mean Time to Recovery (click to expand)</b></summary>

```json showLineNumbers

> {

Reviewer: @kodjomiles to simplify copy-paste maybe we can remove in all these examples of metrics the surrounding curly brace?

I clicked on copy, pasted it under aggregation properties - and then had to manually delete the surrounding braces

   "mean_time_to_recovery": {
      "title": "Mean Time to Recovery",
      "icon": "DefaultProperty",
      "type": "number",
      "target": "pagerdutyIncident",
      "query": {
         "combinator": "and",
         "rules": [
            {
               "property": "status",
               "operator": "=",
               "value": "resolved"
            }
         ]
      },
      "calculationSpec": {
         "func": "average",
         "averageOf": "month",
         "property": "recoveryTime",
         "measureTimeBy": "$createdAt",
         "calculationBy": "property"
      }
   }
}
```

</details>

</TabItem>

</Tabs>

### Complete setup

Here is the complete **Service Blueprint** with the aggregation properties defined for DORA metrics:
<details>

> <summary><b>Complete service blueprint</b></summary>

Reviewer: @kodjomiles I think we discussed that we should put here only the additional properties we want to add, and not all the blueprint. No?

Author: no please, we defined the additional properties and provided the complete setup once we were done to give the client a high overview

Reviewer: @kodjomiles I'm just concerned it will override other stuff the customer added.
Maybe we can remove it completely and just let him go through each metric? what do you think?

Author: okay

---
sidebar_position: 10
displayed_sidebar: null
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import PortTooltip from "/src/components/tooltip/tooltip.jsx";

> # DORA metrics

Reviewer 2: ```suggestion
# Track DORA metrics
```

This guide is designed
to help you

> understand and implement [DevOps Research and Assessment (DORA) metrics](https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance) within your organization in Port.

Reviewer 2: ```suggestion
implement and track [DevOps Research and Assessment (DORA) metrics](https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance) within your organization in Port.  
```

DORA Metrics are a set of key performance indicators
that measure the effectiveness and efficiency of your software development and delivery process. 
By tracking these metrics,
you can identify areas for improvement and ensure that your team is delivering high-quality software efficiently.

> This guide will cover the four key metrics: Deployment Frequency, Lead Time, Change Failure Rate, and Mean Time to Recovery.

Reviewer 2: ```suggestion
This guide will cover the four key metrics: **deployment frequency**, **lead time**, **change failure rate**, and **mean time to recovery**.
```

> :::info Prerequisites

Reviewer 2: Please list the prerequisites under a header (## Prerequisites) and not in an admonition.

- Complete the [Port onboarding process](https://docs.getport.io/quickstart).
- Access a GitHub repository (e.g., port-dora-metrics) to sync your pull requests or deployments. You can follow the [GitHub integration guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github) to ensure your deployments are tracked in Port.
- Optional for advanced strategies: If you're using workflows or pipelines, ensure they are configured for deployment tracking by following the relevant setup guides, such as CI/CD or GitHub Actions.
:::

> :::tip Adding JSON Schema Using Port's UI

Reviewer 2: this tip doesn't make much sense here, it is out of context. Let's add this information where it's needed, for example in the "Data model setup" section it can be added as a last step (click "Edit JSON").

In this setup:
- The `query` checks if the repository is the monorepo and if the file path starts with either `service-A/` or `service-B/`.
- The `identifier` is mapped to the path, which uniquely identifies the service.
- The `title` is derived from the service name in the file path.

By this method, individual services within a monorepo are mapped to Port blueprints.

:::tip Custom Integration Benefits
Custom integrations provide flexibility in mapping and tracking each service or microservice within a monorepo. With Port’s API, you can track deployments and updates for each component separately, giving you granular control over monitoring and managing services in a monorepo.
:::

## Tracking Incidents

Incidents are essential for tracking key DORA metrics, including **Change Failure Rate (CFR)** and **Mean Time to Recovery (MTTR)**. Effective incident tracking reveals insights into how frequently deployments fail and how quickly teams can resolve issues. This section outlines how to:

- Use incidents to calculate **CFR** and **MTTR**.
- Link incidents to services to track the impact of failures.
- Aggregate metrics across incidents for better monitoring.

### Data Model Setup

> Ensure that your **PagerDuty incident blueprint** is configured properly. You can follow the detailed steps in the [PagerDuty Incident Blueprint](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/incident-management/pagerduty).

Reviewer 2: Unclear what these steps are, needs elaborating. Also, the link takes to the head of the page, perhaps you can link directly to the incident header?

Author: I agree.

<summary><b>PagerDuty Incident Relations</b> (click to expand)</summary>

 ```json showLineNumbers
  "relations": {
    "gitHubRepository": {
      "title": "Github Service",
      "target": "service",
      "required": false,
      "many": false
    }
  }

```
</details>

Add the mapping config to pagerduty incident [data source](https://app.getport.io/settings/data-sources):
<details>
<summary><b>Incident mapping config (click to expand)</b></summary>

```yaml showLineNumbers
   gitHubRepository:
   combinator: '"and"'
   rules:
     - property: '"$title"'
       operator: '"="'
       value: .title
```
</details>

:::tip Mapping Incidents to Services
we use the **search relation** entity to map the `pagerdutyIncident` blueprint to the correct service based on the service's `$title` and the pagerduty incident title.
To learn more about using search relations, see [our documentation on Mapping Relations
Using Search Queries](https://docs.getport.io/build-your-software-catalog/customize-integrations/configure-mapping/#mapping-relations-using-search-queries). You can modify this query to map them based on properties of your choice
:::

### Aggregating Incident Data for MTTR

> Add the following **aggregation properties** to the **Incident blueprint**

Reviewer: Incident or Service? I think Service, no?

Reviewer: Actually why is it even required here? don't we cover it in the metrics part?

@kodjomiles

Author: we do but we use this when we want to mirror from the pagerduty service blueprint the mttr aggregated property on the service bp

Author: but since it's not referenced anywhere I might as well remove it

> Add the following **aggregation properties** to the **Incident blueprint**

Reviewer 2: Make sure that things like this are explained at least once. Users will probably not know how to add an aggregation property to a blueprint.

Author: Yeah, I agree.. it was like that initially but we talked about making it very minimalistic. let me add a link that would guide them to how aggregation works

Author: we don't need that section anymore

            "property": "status",
            "operator": "=",
            "value": "resolved"
          }
        ]
      },
      "calculationSpec": {
        "calculationBy": "entities",
        "func": "average",
        "averageOf": "hour",
        "field": "recoveryTime",
        "measureTimeBy": "$createdAt"
      }
    }
  }

```
</details>

This aggregation setup will calculate the **MTTR** by considering incidents that have been resolved,
giving a better picture of recovery performance across services.

## Metrics

We will now aggregate the **DORA metrics** at the **service level**, allowing us to track metrics such as **Deployment Frequency**, **Change Lead Time**, **Change Failure Rate (CFR)**, and **Mean Time to Recovery (MTTR)** for each service.

:::note Metrics Aggregation
If you want to track these metrics at higher levels, such as **team** or **domain**, makes sure the appropriate **team** or **domain** blueprints exist, and that they have relationships defined with the **service** blueprint. Then, you can apply the aggregation properties for these higher hierarchies, similar to how we are doing for the **service** blueprint below.
:::

###  Aggregation

> 

Reviewer: @kodjomiles add here that the timeframes are monthly but you can easily switch to weekly etc.

Author: okay sure thing

{ label: 'Deployment Frequency', value: 'df' },
{ label: 'Change Lead Time', value: 'clt' },
{ label: 'Change Failure Rate', value: 'cfr' },
{ label: 'Mean Time to Recovery', value: 'mttr' }
]}>

<TabItem value="df" label="Deployment Frequency">
Add this aggregation property to calculate deployment frequency:
<details>
<summary><b>Deployment Frequency (click to expand)</b></summary>

```json showLineNumbers
{
  "deployment_frequency": {
    "title": "Monthly Deployment Frequency",
    "icon": "DeploymentIcon",
    "type": "number",
    "target": "deployment",
    "query": {
      "combinator": "and",
      "rules": [
        {
          "property": "deploymentStatus",
          "operator": "=",
          "value": "Success"
        }
      ]
    },
    "calculationSpec": {
      "func": "count",
      "averageOf": "month",
      "measureTimeBy": "$createdAt",
      "calculationBy": "entities"
    }
  }
}
```
</details>
</TabItem>

<TabItem value="clt" label="Change Lead Time">
Add this aggregation property to calculate the lead time for changes:
<details>
<summary><b>Lead Time for Change (click to expand)</b></summary>

```json showLineNumbers
{
  "lead_time_for_change": {

>     "title": "Lead Time for Change",

Author: got it

    "calculationSpec": {
      "func": "average",
      "averageOf": "hour",
      "property": "leadTimeHours",
      "calculationBy": "property",
      "measureTimeBy": "$createdAt"
    }
  }
}
```
</details>
</TabItem>

<TabItem value="cfr" label="Change Failure Rate">
Add the following to the aggregated property in service:
<details>
<summary><b>CFR aggregation property (click to expand)</b></summary>

```json showLineNumbers
{
  "resolved_incidents": {
    "title": "Resolved Incidents",
    "type": "number",
    "target": "pagerdutyIncident",
    "query": {
      "combinator": "and",
      "rules": [
        {
          "property": "status",
          "operator": "=",
          "value": "resolved"
        },
        {
          "property": "incidentType",
          "operator": "=",
          "value": "Deployment"
        }
      ]
    },
    "calculationSpec": {
      "func": "count",
      "calculationBy": "entities"
    }
  },
  "total_deployments": {
    "title": "Total Deployments",
    "type": "number",
    "target": "deployment",
    "query": {
      "combinator": "and",
      "rules": [
        {
          "property": "deploymentStatus",
          "operator": "=",

>           "value": "Success"

Reviewer: @kodjomiles why aren't we showing here timerframe?
e.g. monthly in our example

Aggregate per month

Author: we are getting the total value for all deployments so I thought that should  not be  based on a timeframe

Reviewer: I think we should follow the guideline of monthly metrics
What you did shows the rate lifetime. Instead we want to measure the rate on a monthly basis. WDYT?

Author: Yeah sounds good

        }
      ]
    },
    "calculationSpec": {
      "func": "count",
      "calculationBy": "entities"
    }
  }
}
```
</details>

Add this calculation property to calculate the cfr from the aggregated properties:
<details>
<summary><b>CFR calculation property (click to expand)</b></summary>
    
```json showLineNumbers
    {
      "changeFailureRate": {
        "title": "Change Failure Rate",
        "calculation": "(.properties.resolved_incidents / .properties.total_deployments) * 100",
        "type": "number"
      }
    }
     
```
</details>

</TabItem>

<TabItem value="mttr" label="Mean Time to Recovery">
Add this aggregation property to calculate the MTTR:

<details>
<summary><b>Mean Time to Recovery (click to expand)</b></summary>

```json showLineNumbers
{
  "mean_time_to_recovery": {
    "title": "Mean Time to Recovery",
    "type": "number",
    "target": "pagerdutyIncident",
    "query": {
      "combinator": "and",
      "rules": [
        {
          "property": "status",
          "operator": "=",
          "value": "resolved"
        }
      ]
    },

>     "calculationSpec": {

Reviewer: @kodjomiles why not monthly?

            "value": "resolved"
          },
          {
            "property": "incidentType",
            "operator": "=",
            "value": "Deployment"
          }
        ]
      },
      "calculationSpec": {
        "func": "count",
        "calculationBy": "entities"
      }
    },
    "total_deployments": {
      "title": "Total Deployments",
      "type": "number",
      "target": "deployment",
      "query": {
        "combinator": "and",
        "rules": [
          {
            "property": "deploymentStatus",
            "operator": "=",
            "value": "Success"
          }
        ]
      },
      "calculationSpec": {
        "func": "count",
        "calculationBy": "entities"
      }
    }
  },
  "relations": {
    "admins": {
      "title": "Admins",
      "target": "githubUser",
      "required": false,
      "many": true
    },
    "githubTeams": {
      "title": "GitHub Teams",
      "target": "githubTeam",
      "required": false,
      "many": true
    }
  }
}
```
</details>

## Visualization
By leveraging Port's Dashboards, you can create custom dashboards to track the metrics and monitor your team's performance over time.

> ### Dashboard Setup

Reviewer 2: make sure to be specific with links and instructions, and put button names in `quotes like this` (See my suggestions on points 1 and 2)

> 1. Go to **Catalog** in the navigation bar

Reviewer 2: ```suggestion
1. Go to your [software catalog](https://app.getport.io/organization/catalog).
```

> 2. Click **new**

Reviewer 2: ```suggestion
2. Click on the `+ New` button in the left sidebar.
```

1. Click **+ Widget** and select **Number Chart**
2. Title: `Deployment Frequency - Monthly`, (add the rocket icon)
3. Select `Display single property` and choose **Service** as the **Blueprint**
4. Select an `Entity` and choose `Monthly Deployment Frequency` as the **Property**

   <img src="/img/guides/deploymentFrequencyChartDoraMetrics.png" width="50%"/>
   
5. Click **Save**

</details>

<details>
<summary><b>Setup MTTR Widget</b></summary>

1. Click **+ Widget** and select **Number Chart**
2. Title: `MTTR – Monthly Average (Seconds)`, (add the pagerduty icon)
3. Select `Display single property` and choose **Service** as the **Blueprint**
4. Select an `Entity` and choose `Mean Time to Recovery` as the **Property**

   <img src="/img/guides/mttrDoraMetricsChart.png" width="50%"/>
   
5. Click **Save**

</details>

<details>
<summary><b>Setup Change Lead Time Widget</b></summary>

1. Click **+ Widget** and select **Number Chart**
2. Title: `Lead Time for Changes (Hours)`, (add LineChart icon)
3. Select `Display single property` and choose **Service** as the **Blueprint**
4. Select an `Entity` and choose `Lead Time for Change` as the **Property**

> 5. Select the `Hour` time interval and preferred time range

Reviewer: Why not monthly? 
@kodjomiles

2. Title: `Change Lead Time Over Time`, (add the LineChart icon)
3. Choose **Service** as the **Blueprint**
4. Select an `Entity` and choose `Lead time for change` as the **Property**
5. Select the preferred time interval

   <img src="/img/guides/changeLeadTimeOverTimeDM.png"  width="50%"/>

6. Click **Save**.

</details>

<details>
<summary><b>Deployments Frequency Over Time</b></summary>

1. Click **+ Widget** and select **Line Chart**
2. Title: `Deployments Frequency Over Time` (add the rocket icon)
3. Select **Service** as the **Blueprint**
4. Select `Monthly Deployment Frequency` as the **Property**
5. Set **Time Interval** to `Month` and **Time Range** to `In the past 365 days`

   <img src="/img/guides/deploymentFrequencyOverTime.png" width="50%"/>

6. Click **Save**.
</details>

:::tip Metric widget groupings
It would be visually cleaner and more informative to group related widgets, such as the **Line Chart** and **Number Chart** widgets, side by side for easier comparison.
You can replicate more examples by checking our dora metrics dashboard on the [demo environment](https://demo.getport.io/dora_metrics).
:::

<img src="/img/guides/doraMetricsDBVisualization.png"/>

> By following these steps, you can create a custom dashboard to visualize the DORA metrics.

Reviewer 2: ```suggestion
```

> Congrats 🎉 You've successfully setup DORA metrics in Port 🔥

Reviewer 2: ```suggestion
Congrats 🎉 You have successfully set up DORA metrics tracking in your portal 🔥
```
